### PR TITLE
fix(userspace/libsinsp): solve cases of unsafe string_view access

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind bug

**Any specific area of the project related to this PR?**

/area libsinsp

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

While bumping the libs dependency in Falco I noticed a bunch of rightful complaints about unsafe OOB access to string views elements.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

/milestone 0.17.0

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by FedeDP — DISMISSED** [review]
/approve

File: userspace/libsinsp/parsers.cpp [hunk 1]
```diff
@@ -2496,7 +2496,7 @@ std::string sinsp_parser::parse_dirfd(sinsp_evt *evt, std::string_view name, int
 	/* This should never happen but just to be sure. */
 	if(name.data() != nullptr)
 	{
-		is_absolute = (name[0] == '/');
+		is_absolute = (name.size() > 0 && name[0] == '/');
```
Comment (federico-sysdig) [inline userspace/libsinsp/parsers.cpp:2499]: `empty()` has a chance of being more efficient. I'd use that when testing for zero size.
 
```suggestion
		is_absolute = (!name.empty() && name[0] == '/');
```

File: userspace/libsinsp/sinsp_filtercheck_k8s.cpp [hunk 2]
```diff
@@ -141,16 +141,16 @@ int32_t sinsp_filter_check_k8s::extract_arg(string_view fldname, string_view val
...
-			if(!isalnum(val[endpos])
-				&& val[endpos] != '/'
-				&& val[endpos] != '_'
-				&& val[endpos] != '-'
-				&& val[endpos] != '.')
+			if(!isalnum(val.at(endpos))
+				&& val.at(endpos) != '/'
+				&& val.at(endpos) != '_'
+				&& val.at(endpos) != '-'
+				&& val.at(endpos) != '.')
```
Comment (federico-sysdig) [inline userspace/libsinsp/sinsp_filtercheck_k8s.cpp:153]: Why favoring `at()` over `operator[]`? The first performs bound checking, but in this case we are iterating inside the string `val`, so the bound checking is an extra calculation we could avoid.
Actually, all uses of `at()` in the PR seem to be after a manual bound check, so they could all be reverted to index operation without harm.

**Review by federico-sysdig — COMMENTED** [review]
Good catch! I noticed that the replacements of index operations (`operator[]`) with `at()` occur in places where we are already ensuring the index is inside the string. I think we should avoid `at()` as it adds an extra bound check (and throws an exception in case).

**Comment by jasondellaluce** [comment]
> Good catch! I noticed that the replacements of index operations (`operator[]`) with `at()` occur in places where we are already ensuring the index is inside the string. I think we should avoid `at()` as it adds an extra bound check (and throws an exception in case).

This is true, however using `at()` will be less error probe. My goal was not to achieve corrected, since as you said we're basically checking things twice, but rather making the code resilient for future changes whenever the next person will touch this code in the future. `at()` adds an extra layer of protection at very little cost. Consider that these functions are only invoked at load/reload time and not in the hot path, so there's really no performance implication compared to `[]`.

**Review by FedeDP — APPROVED** [review]
/approve

**Review by incertum — APPROVED** [review]
/approve

**Pull request merged**